File: 08. Trees/07. Diameter of a binary tree/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

function diameterOfBinaryTree(root) {
    let diameter = 0;

    function depth(node) {
        if (!node) return 0;

        let leftDepth = depth(node.left);
        let rightDepth = depth(node.right);

        diameter = Math.max(diameter, leftDepth + rightDepth);

        return 1 + Math.max(leftDepth, rightDepth);
    }

    depth(root);
    return diameter;
}

// Main function
function main() {
    let root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.left = new TreeNode(4);
    root.left.right = new TreeNode(5);

    console.log(diameterOfBinaryTree(root)); // Output: 3
}

// Run the main function
main();

File: 08. Trees/07. Diameter of a binary tree/main.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


int diameter = 0;

int depth(TreeNode* root) {
    if (!root) return 0;

    int leftDepth = depth(root->left);
    int rightDepth = depth(root->right);

    diameter = max(diameter, leftDepth + rightDepth);

    return 1 + max(leftDepth, rightDepth);
}

int diameterOfBinaryTree(TreeNode* root) {
    depth(root);
    return diameter;
}


int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    cout << diameterOfBinaryTree(root) << endl;  // Output: 3

    return 0;
}
